pkg/compactor/shuffle_sharding_grouper.go

```diff
@@ -142,103 +157,183 @@ func (g *ShuffleShardingGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (re
+		break mainLoop
+	}
+
+	level.Info(g.logger).Log("msg", fmt.Sprintf("total groups for compaction: %d", len(outGroups)))
+
 	return outGroups, nil
 }
 
-// Check whether this compactor instance owns the group.
-func (g *ShuffleShardingGrouper) ownGroup(groupHash uint32) (bool, error) {
-	subRing := g.ring.ShuffleShard(g.userID, g.limits.CompactorTenantShardSize(g.userID))
-
-	rs, err := subRing.Get(groupHash, RingOp, nil, nil, nil)
-	if err != nil {
-		return false, err
+func (g *ShuffleShardingGrouper) isGroupLocked(blocks []*metadata.Meta) (bool, error) {
+	for _, block := range blocks {
+		blockID := block.ULID.String()
+		lockFileReader, err := g.bkt.Get(g.ctx, path.Join(blockID, BlockLockFile))
+		if err != nil {
+			if g.bkt.IsObjNotFoundErr(err) {
+				level.Debug(g.logger).Log("msg", fmt.Sprintf("no lock file for block: %s", blockID))
+				continue
+			}
+			return true, err
+		}
+		b, err := ioutil.ReadAll(lockFileReader)
+		if err != nil {
+			level.Error(g.logger).Log("msg", fmt.Sprintf("unable to read lock file for block: %s", blockID), "err", err)
+			return true, err
+		}
+		blockLocker := BlockLocker{}
+		err = json.Unmarshal(b, &blockLocker)
+		if err != nil {
+			level.Error(g.logger).Log("msg", fmt.Sprintf("unable to parse lock file for block: %s", blockID), "err", err)
+			return true, err
+		}
+		if time.Now().Before(blockLocker.LockTime.Add(HeartBeatTimeout)) {
+			level.Debug(g.logger).Log("msg", fmt.Sprintf("locked block: %s", blockID))
+			return true, nil
+		}
 	}
+	return false, nil
+}
 
-	if len(rs.Instances) != 1 {
-		return false, fmt.Errorf("unexpected number of compactors in the shard (expected 1, got %d)", len(rs.Instances))
+func (g *ShuffleShardingGrouper) groupLockHeartBeat(blocks []*metadata.Meta) {
+	var blockIds []string
+	for _, block := range blocks {
+		blockIds = append(blockIds, block.ULID.String())
 	}
+	blocksInfo := strings.Join(blockIds, ",")
+	level.Info(g.logger).Log("msg", fmt.Sprintf("start heart beat for blocks: %s", blocksInfo))
+heartBeat:
+	for {
+		select {
+		case <-g.ctx.Done():
+			break heartBeat
+		default:
+			level.Debug(g.logger).Log("msg", fmt.Sprintf("heart beat for blocks: %s", blocksInfo))
+			g.lockBlocks(blocks)
+			time.Sleep(HeartBeatTimeout / 5) // it allows up to 5 failures on heart heat for single block
+		}
+	}
+	level.Info(g.logger).Log("msg", fmt.Sprintf("stop heart beat for blocks: %s", blocksInfo))
+}
 
-	return rs.Instances[0].Addr == g.ringLifecyclerAddr, nil
+func (g *ShuffleShardingGrouper) lockBlocks(blocks []*metadata.Meta) {
+	for _, block := range blocks {
+		blockID := block.ULID.String()
+		blockLockFilePath := path.Join(blockID, BlockLockFile)
+		blockLocker := BlockLocker{
+			CompactorID: g.ringLifecyclerID,
+			LockTime:    time.Now(),
+		}
+		lockFileContent, err := json.Marshal(blockLocker)
+		if err != nil {
+			level.Error(g.logger).Log("msg", "unable to create lock file content for block", "block_id", blockID, "err", err)
```

Comment:
could we have a metric to tell us when some operation involving the lock file fails?